drafts/inlinemacros.md

```diff
@@ -165,99 +170,66 @@ positions:
+(in that case, expansion will take place immediately at the place where the macro expression is written).
+
+Macro bodies can only reference the following names in their environment:
+
+ 1. type parameters of enclosing inline methods and classes/traits containing an enclosing inline method as a member,
+ 2. terms parameters of enclosing inline methods (where the parameter may, or may not be, marked inline),
+ 3. `this` references to classes/traits containing an enclosing inline method as a member,
+ 4. inline values,
+ 5. inline methods,
+ 6. anything that's global (i.e. a class/trait/object without an outer reference).
+
+Names referenced in macro bodies undergo certain changes. References to inline and global definitions are left untouched,
+but local terms of type `T` are transformed into terms of type `scala.meta.Term[T]`, 
+and local types `U` are transformed into terms of type `scala.meta.Type`. In other words, 
+definitions that are statically available outside macro bodies remain available in macro bodies,
+whereas term and type parameters of enclosing inline methods become available as their representations.
+
+## Macro expansion
+
+During typechecking, the compiler treats macro expressions, i.e. invocations of `Predef.macro`,
+as normal method calls, typechecking macro bodies and performing type inference if necessary, but nothing else.
+In this proposal, macro expansion works differently from macro expansion in Scala 2.10+, 
+where macro applications are expanded immediately after being processed by the typechecker.
+
+An important consequence is that macro expressions cannot refine their types during expansion.
+This makes it impossible to use the proposed macro system to implement [whitebox macros](http://docs.scala-lang.org/overviews/macros/blackbox-whitebox.html) from Scala 2.10+,
+but we are planning to eventually provide alternative ways of enabling the most important whitebox functionality.
+
+At some point in the compilation pipeline, after typechecking is complete for the entirety of the program, 
+the compiler expands macro expressions outside the bodies of inline values and inline methods.
+When macro expressions code appear inside the aforementioned constructs, 
+they will be ignored by the macro expander and processed by the compiler as usual.
+
+A macro expression is expanded by evaluating its body and replacing the original expression
+with an expression that represents the result of the evaluation. 
+The implementation is responsible for instantiating a `scala.meta.macros.Context` necessary for macro bodies
+to evaluate and for converting between its internal representation for program elements and representations
+defined in `scala.meta`, such as `scala.meta.Term` and `scala.meta.Type`.
+
+After the expansion is complete, the result of the expansion is forcibly upcast
+to the type of the original expression. This mechanism works exactly the same 
+as the mechanism of blackbox macro expansion in Scala 2.10+.
```

Comment:
This formulation is almost a verbatim copy of a paragraph describing the same thing about inline expansion. Can we deduplicate somehow?